### ISSUE TITLE
Function to return (cached) cacerts?

### ISSUE THREAD

**Issue by chulkilee** [elixir-mint/castore#22]
From https://github.com/elixir-mint/mint/issues/274

Is there any specific reason this library returns cached cacerts upon request, instead of returning the cafile path?

If we add a function to castore (e.g. `CAStore.cacerts/0`), then we don't need to implement cache feature on other library  individually (such as https://github.com/elixir-mint/mint/pull/275 )

**Comment by ericmj** [comment]
> Is there any specific reason this library returns cached cacerts upon request, instead of returning the cafile path?

It is the other way around. This library has a single function that returns the file path. We need to return the path because that’s the only way to get caching in erlang since they can’t cache cacerts and our own caching wont help since they do additional transformations.

There are many different ways to do caching and invalidation. You can also get different performance depending on what OTP version you are targeting. We can also see from the PR that it’s straightforward to add caching on top.

So I am hesitant to add caching and am happy to keep this library simple.

**Comment by chulkilee** [comment]
Oh, if erlang keeps the cache the contents of cacertfile based on path.. then I agree with you in that we can leave it to erlang.

I found it's called as "PEM cache" in erlang ssl doc.

Closing the issue as this is non-goal of this package.

---

BTW I found this is interesting [Heroku - How We Sped up SNI TLS Handshakes by 5x (2016)](https://blog.heroku.com/how-we-sped-up-sni-tls-handshakes-by-5x) as Heroku's usecase required to disable the erlang cache to leverage preprocessed certificates. Just for future reader...

> The funny bit there is that the SSL library supports an entire other format to handle certificates: have the user decode the PEM data to the DER format themselves and then submit the DER data directly in memory through the connection configuration. For dynamic routers such as Heroku's, that's the method we've taken to avoid storing multiple certificates on disk unencrypted. We store them ourselves, using appropriate cryptographic means, and then decode them only once they are requested through SNI, at which point they are passed to the SSL library.
>
> For this use case, the SSL library has certificates passed straight in memory and does not require file access. Still, the same caching code path was used. The certificates are cached but not reference-counted and also not shared across connections either. For heavy usage of DER-encoded certificates, the PEM cache, therefore, becomes a central bottleneck for the entire server, forcing the decoding of every certificate individually through that single critical process.
